homeassistant/auth/mfa_modules/notify.py

```diff
@@ -0,0 +1,364 @@
+"""HMAC-based One-time Password auth module.
+
+Sending HOTP through notify service
+"""
+import logging
+from collections import OrderedDict
+from random import SystemRandom
+from typing import Any, Dict, Optional, Tuple, List  # noqa: F401
+
+import attr
+import voluptuous as vol
+
+from homeassistant.const import CONF_EXCLUDE, CONF_INCLUDE
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers import config_validation as cv
+
+from . import MultiFactorAuthModule, MULTI_FACTOR_AUTH_MODULES, \
+    MULTI_FACTOR_AUTH_MODULE_SCHEMA, SetupFlow
+
+REQUIREMENTS = ['pyotp==2.2.6']
+
+CONF_MESSAGE = 'message'
+
+CONFIG_SCHEMA = MULTI_FACTOR_AUTH_MODULE_SCHEMA.extend({
+    vol.Optional(CONF_INCLUDE): vol.All(cv.ensure_list, [cv.string]),
+    vol.Optional(CONF_EXCLUDE): vol.All(cv.ensure_list, [cv.string]),
+    vol.Optional(CONF_MESSAGE, default='Your Home Assistant One-time Password'
+                                       ' is "{}"'): str
+}, extra=vol.PREVENT_EXTRA)
+
+STORAGE_VERSION = 1
+STORAGE_KEY = 'auth_module.notify'
+STORAGE_USERS = 'users'
+STORAGE_USER_ID = 'user_id'
+STORAGE_OTA_SECRET = 'ota_secret'
+STORAGE_COUNTER = 'counter'
+
+INPUT_FIELD_CODE = 'code'
+
+DUMMY_SECRET = '7Z5EFWI4RFLVV67G'
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@attr.s(slots=True)
+class NotifySetting:
+    """Store notify setting for one user."""
+
+    secret = attr.ib(type=str)
+    counter = attr.ib(type=int)
+    notify_service = attr.ib(type=Optional[str], default=None)
+    target = attr.ib(type=Optional[str], default=None)
+
+
+_UsersDict = Dict[str, NotifySetting]
+
+
+def _generate_secret_and_init_counter() -> Tuple[str, int]:
+    """Generate a secret and a random initial counter."""
+    import pyotp
+
+    ota_secret = pyotp.random_base32()
+    counter = SystemRandom().randint(0, 2 << 31)
```

Comment:
Can't we just always use a random input for counter value?

Reply:
Also, do we really need to use pyotp to generate a random code between 100000 and 999999 ? What does it do that makes it more secure?

Reply:
For the random counter, strength is same, pyotp internal use `SystemRandom` as well. However I will change it to use pyotp just looks consistent.

pyotp implemented RFC 4226 for the one-time password generation, I thought it will be secure that we write a routine by ourselves.

Reply:
We can use 8 digit number for the code, looks better than 6 digit

Reply:
I prefer 6 digits because it's easier to remember when you are typing it from your phone.